### DISCUSSION TITLE
column_property leads to InvalidRequestError

### DISCUSSION THREAD

**Discussion by finswimmer** [sqlalchemy/sqlalchemy#6765] [Usage Questions]
Hello,

please note that I reported this first on stackoverflow: https://stackoverflow.com/questions/68363025/column-property-leads-to-invalidrequesterror

Because I didn't receive any reaction there and couldn't find any similar case reported elsewhere, I decided to post this here has an issue.

**Describe the bug**

Let's assume we have three tables, which are represented by this sqlalchemy orm models:

* A
* B
* C

A and C have foreign keys to B.

I like to add two relationships to A:

* `b_rel` which is a relationship to `B`
* `c_rel` which is relationship to those `C` which references `B`

Furthermore `B` has a column `name` which I want to use as a property in `C` as well. Doing this by adding a `hybrid_property` to `C` works fine. Using a `column_property` leas to this error:

```
sqlalchemy.exc.InvalidRequestError: Select statement '<sqlalchemy.sql.selectable.Select object at 0x7fbb151dff70>' returned no FROM clauses due to auto-correlation; specify correlate(<tables>) to control correlation manually.
```

**To Reproduce**

**model.py**
```py
from sqlalchemy import Column, ForeignKey, Integer, String, select
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import column_property, relationship

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    b_id = Column(ForeignKey("b.id"))

    b_rel = relationship("B")
    c_rel = relationship("C", secondary="join(B, C)", primaryjoin="A.b_id == B.id")


class B(Base):
    __tablename__ = "b"

    id = Column(Integer, primary_key=True)
    name = Column(String)


class C(Base):
    __tablename__ = "c"

    id = Column(Integer, primary_key=True)
    b_id = Column(ForeignKey("b.id"))

    # Using column_property leads to:
    #
    # sqlalchemy.exc.InvalidRequestError: Select statement '<sqlalchemy.sql.selectable.Select object at 0x7fbb151dff70>'
    # returned no FROM clauses due to auto-correlation; specify correlate(<tables>) to control correlation manually.
    b_name = column_property(select([B.name]).where(B.id == b_id).scalar_subquery())

    # Using relationship together with hybrid_property works
    #
    # b_ = relationship("B")
    #
    # @hybrid_property
    # def b_name(self) -> str:
    #     return self.b_.name
    #
    # @b_name.expression
    # def b_name(self) -> str:
    #     return (
    #         select([B.name]).select_from(B).where(B.id == self.b_id).scalar_subquery()
    #     )
```

**test_model.py**

```py
def test_relations():
    import model

    if Path("test_database.db").exists():
        Path("test_database.db").unlink()

    engine = create_engine("sqlite:///test_database.db", echo="debug")

    model.Base.metadata.create_all(engine)
    session = sqlalchemy.orm.Session(autocommit=False, autoflush=False, bind=engine)

    b = model.B(name="test")
    session.add(b)
    session.commit()

    a = model.A(b_id=b.id)
    c = model.C(b_id=b.id)
    session.add_all([a, c])
    session.commit()

    assert a.b_rel == b
    assert a.c_rel[0] == c

    c_search = session.query(model.C).filter_by(b_name="test").one()
    assert c_search == c
```

**Error**
```
sqlalchemy.exc.InvalidRequestError: Select statement '<sqlalchemy.sql.selectable.Select object at 0x7fbb151dff70>' returned no FROM clauses due to auto-correlation; specify correlate(<tables>) to control correlation manually.
```

**Versions.**
 - OS: Ubuntu 18.04, ArchLinux
 - Python: python3.7 - python3.9
 - SQLAlchemy: 1.4.20

**Comment by zzzeek** [comment]
the error message:

" Select statement '<sqlalchemy.sql.selectable.Select object at 0x7fbb151dff70>' returned no FROM clauses due to auto-correlation; specify correlate(<tables>) to control correlation manually."

the "correlate" it's talking about is [this](https://docs.sqlalchemy.org/en/14/core/selectable.html?highlight=correlate#sqlalchemy.sql.expression.ScalarSelect.correlate) which is more generally discussed at https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#tutorial-scalar-subquery.     In this case , you want to correlate "C" and any alias of "C", so it's easier to use correlate_except(B):

    b_name = column_property(
        select([B.name]).where(B.id == b_id).correlate_except(B).scalar_subquery()
    )

this is basically the same thing you'll see in the example for ORM column_property with correlated subqueries [here](https://docs.sqlalchemy.org/en/14/orm/mapped_sql_expr.html#using-column-property).      hope this helps

**Reply by finswimmer** [reply]
Thanks a lot for your quick answer @zzzeek :+1: 

I don't fully understand what's going on but at least an idea. I guess I have to reread the linked sources several times :)

fin swimmer
